### PR TITLE
schemacmp: support partition

### PR DESCRIPTION
**Pull request by sleepymole**
### What problem does this PR solve? 

Support comparing and joining table with partition in schemacmp.

### What is changed and how it works?

Encode partition info into `schemacmp.Table` and compare it.

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

### CODE REVIEW

File: pkg/schemacmp/lattice.go [hunk 1]
```diff
@@ -469,6 +469,42 @@ func (a Uint) Join(other Lattice) (Lattice, error) {
 	}
 }
 
+// Uint64 is a uint implementing Lattice.
+type Uint64 uint
```
Comment (lance6716) [inline pkg/schemacmp/lattice.go]: uint64 🤔

File: pkg/schemacmp/table.go [hunk 2]
```diff
@@ -235,6 +235,168 @@ func (indexMap) JoinWithNil(value Lattice) (Lattice, error) {
...
+	num := Singleton(pi.Num)
+	// Only range and list partition num can be changed.
+	if pi.Type == model.PartitionTypeRange || pi.Type == model.PartitionTypeList {
+		num = Uint64(pi.Num)
+	}
+	return Tuple{
+		Singleton(pi.Type),
+		Singleton(pi.Expr),
+		columns,
+		Singleton(pi.Enable),
```
Comment (lance6716) [inline pkg/schemacmp/table.go]: by reading the comment of `pi.Enable`, I think this is a TiDB specified attribute so we don't need it.

**Review by lance6716 — COMMENTED** [review]
what's the behaviour for https://dev.mysql.com/doc/refman/5.7/en/partitioning-subpartitions.html

**Comment by sleepymole** [comment]
> what's the behaviour for https://dev.mysql.com/doc/refman/5.7/en/partitioning-subpartitions.html

I didn't find subpartitioning in model.TableInfo. I will check it again later.

File: pkg/schemacmp/lattice.go (hunk 1 above)
Comment (sleepymole) [reply pkg/schemacmp/lattice.go]: thanks, this is a typo

File: pkg/schemacmp/table.go [hunk 3]
```diff
@@ -235,6 +235,168 @@ func (indexMap) JoinWithNil(value Lattice) (Lattice, error) {
...
+		ctx.WritePlain(" (")
+		for i, pair := range sortedMap(info[partitionTupleIndexDefinitions].(map[string]interface{})) {
+			if i > 0 {
+				ctx.WritePlain(", ")
+			}
+			ctx.WriteKeyWord("PARTITION ")
+			ctx.WriteName(pair.key)
+
+			d := pair.value.(partitionDefinition)
+			if len(d.lessThan) > 0 {
```
Comment (lance6716) [inline pkg/schemacmp/table.go:382]: in create table AST, value less than could be an expression. So here did it evaluate and only save the result?

File: pkg/schemacmp/table_test.go [hunk 4]
```diff
@@ -518,20 +608,54 @@ func (s *tableSchema) TestJoinSchemas(c *C) {
...
 		}
 	}
+
+	// Table with partition should be parsed after encoded.
+
+	// Range partition.
+	ti, err = s.toTableInfo("CREATE TABLE tb (id INT NOT NULL) PARTITION BY RANGE (id) (PARTITION p0 VALUES LESS THAN (10), PARTITION p1 VALUES LESS THAN (20))")
+	c.Assert(err, IsNil)
+	sql := strings.ToLower(Encode(ti).String())
+	c.Assert(strings.Contains(sql, "partition by range"), Equals, true)
```
Comment (lance6716) [inline pkg/schemacmp/table_test.go:635]: I think above line we use ToLower to handle `charset` and `CHARSET`, here no need to use it

File: pkg/schemacmp/table.go (hunk 3 above)
Comment (sleepymole) [reply pkg/schemacmp/table.go:382]: yes, table info only saved the string results.

File: pkg/schemacmp/table_test.go (hunk 4 above)
Comment (sleepymole) [reply pkg/schemacmp/table_test.go:635]: here is to check whether the sql  statement contains the partition definition.

**Comment by lance6716** [comment]
we should also forbid "using ADD to rollback a partitially dropped partition" 😭 
and REORGANIZE is different from ADD, we might permit reorg? 😵 

 https://dev.mysql.com/doc/refman/5.7/en/partitioning-management-range-list.html

File: pkg/schemacmp/table_test.go (hunk 4 above)
Comment (lance6716) [reply pkg/schemacmp/table_test.go:635]: We could remove ToLower and check PARTITION BY RANGE

**Comment by sleepymole** [comment]
> we should also forbid "using ADD to rollback a partially dropped partition" 😭

If we just remove a partition, I think the upstream should make sure all data in the partition should be deleted. And there is a `TRUNCATE PARTITION`  operation which is similar with `DROP COLUMN`.  But it seems `TRUNCATE` doesn't change the table structure. So we can't do much in schemacmp.

> and REORGANIZE is different from ADD, we might permit reorg? 😵

TiDB doesn't support reorg currently. And there are more operations which are not supported by TiDB. If the downstream is always TiDB, I think we can forbid these operations.
> With regard to partition management, any operation that requires moving data in the bottom implementation is not supported currently, including but not limited to: adjust the number of partitions in a Hash partitioned table, modify the Range of a Range partitioned table, merge partitions and exchange partitions.

File: pkg/schemacmp/table_test.go (hunk 4 above)
Comment (sleepymole) [reply pkg/schemacmp/table_test.go:635]: `charset` check is no need to use `ToLower` too 😂

Comment (lance6716) [reply pkg/schemacmp/table_test.go:635]: maybe the original auther wants to check `charset` and `CHARSET`

Comment (sleepymole) [reply pkg/schemacmp/table_test.go:635]: I'm the original author 😂

Comment (sleepymole) [reply pkg/schemacmp/table_test.go:635]: I originally wanted to increase the stability, because the SQL keywords are case-insensitive.

**Comment by sleepymole** [comment]
> > what's the behaviour for https://dev.mysql.com/doc/refman/5.7/en/partitioning-subpartitions.html
> 
> I didn't find subpartitioning in model.TableInfo. I will check it again later.

@lance6716  Subpartitioning are only supported in parser. TiDB doesn't support subpartitioning  currently and there is no subpartitioning  info in `model.TableInfo`.

File: pkg/schemacmp/table.go [hunk 5]
```diff
@@ -235,6 +235,168 @@ func (indexMap) JoinWithNil(value Lattice) (Lattice, error) {
...
+		av := a.inValues[i]
+		bv := b.inValues[i]
+		if len(av) != len(bv) {
+			return false
+		}
+		for j := 0; j < len(av); j++ {
+			if av[j] != bv[j] {
+				return false
+			}
+		}
```
Comment (lichunzhu) [inline pkg/schemacmp/table.go:269]: We can write a sub-function for this part and L248-255

**Comment by lance6716** [comment]
Now we know that DROP PARTITION will delete the data in that PARTITION.

for two sharding tables t1, t2 with same partition [MIN_VALUE, 100), [100, 200). consider this case:

1. t1 DROP PARTITION [MIN_VALUE, 100), now the only partition is [MIN_VALUE, 200). the joined didn't change, so no DDL executed
2. t1 INSERT 50
3. t2 DROP PARTITION [MIN_VALUE, 100). the joined changed, downstream DROP PARTITION [MIN_VALUE, 100)

but downstream will lost INSERT 50 in step 2 🤔 maybe optimistic sharding can't support PARTITION

**Comment by sleepymole** [comment]
> Now we know that DROP PARTITION will delete the data in that PARTITION.
> 
> for two sharding tables t1, t2 with same partition [MIN_VALUE, 100), [100, 200). consider this case:
> 
> 1. t1 DROP PARTITION [MIN_VALUE, 100), now the only partition is [MIN_VALUE, 200). the joined didn't change, so no DDL executed
> 2. t1 INSERT 50
> 3. t2 DROP PARTITION [MIN_VALUE, 100). the joined changed, downstream DROP PARTITION [MIN_VALUE, 100)
> 
> but downstream will lost INSERT 50 in step 2 🤔 maybe optimistic sharding can't support PARTITION

too sad ...

**Comment by sleepymole** [comment]
As the discussion above, it seems there is no way that optimistic sharding will support PARTITION. @lance6716  How about close this PR.

**Comment by lance6716** [comment]
(closing, thanks)

**Pull request closed without merging**